src/backend/distributed/metadata/node_metadata.c

```diff
@@ -1053,17 +1054,26 @@ RemoveNodeFromCluster(char *nodeName, int32 nodePort)
 	WorkerNode *workerNode = ModifiableWorkerNode(nodeName, nodePort);
 	if (NodeIsPrimary(workerNode))
 	{
+		if (CanRemoveReferenceTablePlacements())
+		{
+			/*
+			 * Delete reference table placements so they are not taken into account
+			 * for the check if there are placements after this.
+			 */
+			DeleteAllReferenceTablePlacementsFromNodeGroup(workerNode->groupId);
+		}
 		bool onlyConsiderActivePlacements = false;
-
-		/*
-		 * Delete reference table placements so they are not taken into account
-		 * for the check if there are placements after this
-		 */
-		DeleteAllReferenceTablePlacementsFromNodeGroup(workerNode->groupId);
-
 		if (NodeGroupHasShardPlacements(workerNode->groupId,
 										onlyConsiderActivePlacements))
 		{
+			if (ClusterHasReferenceTable())
+			{
+				ereport(ERROR, (errmsg(
+									"you cannot remove node because there are reference "
+									"tables and you would lose the data on reference tables"),
+								errhint(
+									"drop all reference tables before removing the node")));
+			}
 			ereport(ERROR, (errmsg("you cannot remove the primary node of a node group "
```

Comment:
We  have this question on slack right now. Can we now improve `you cannot remove the primary node of a node group which has shard placements` to have an hint to drop distributed tables?

Reply:
Actually should we mention `undistribute_table` here and also in reference table case? Because if people drop the table then they lose the data :)